[huey.j 전준호] subway-service step3 코드 리뷰 부탁드립니다.

jh280722 opened the pull request:
안녕하세요 휴이입니다.
이번 지하철 노선도 step3 코드 리뷰 부탁드립니다.
하면서 인증에 대한 부분을 어떻게 처리해야하는지 고민을 많이 했는데요
path부분에서 로그인을 하지않아도 결과를 얻을 수 있고 로그인을 하면 나이에 따라 할인이 되어 출력이 되길래
resolveArgument에서 인증 토큰이 없을 경우는 null을 리턴하여 처리하였습니다.
이 부분이 맞게 한 것인지 궁금하고요.
도메인의 서비스에서 다른 도메인의 dao를 이용하고 싶을때 서비스를 불러와서 처리하는지 dao를 불러와서 처리하는지가 궁금합니다.
감사합니다!

Commit:
feat : 최단 거리 경로 조회 구현

Commit:
refactor : 오타 수정

Commit:
feat : auth Exception 추가 및 Member crup 권한 추가

Commit:
feat : 즐겨찾기 기능 구현

Commit:
feat : extraFare 계산 기능 구현 및 리펙토링

limmmee-bonjour commented on src/main/java/subway/AuthenticationPrincipalConfig.java:
```diff
@@ -12,6 +13,7 @@
 public class AuthenticationPrincipalConfig implements WebMvcConfigurer {
     private final AuthService authService;
 
+    @Autowired
```
DI 시 autowired 어노테이션과 생성자 주입을 함께 쓰는 이유가 궁금합니다 !
[이 글](https://yaboong.github.io/spring/2019/08/29/why-field-injection-is-bad/)을 읽어보시면 도움이 될거예요 😄

jh280722 replied:
아하 감사합니다. 생성자 주입을 해도 명시적으로 붙여주면 좋을줄 알았습니다.

limmmee-bonjour commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,40 @@
+
+    public boolean checkInvalidMember(String email, String password) {
+        Member member = memberDao.findByEmail(email);
+        if (isMember(password, member)) {
+            return false;
+        }
+        return true;
```
간결하게 한줄로 나타낼 수 있지 않을까요? 😉

jh280722 replied:
아 넵 is니까 그냥 리턴해주면 되겠네요

limmmee-bonjour commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,40 @@
+    }
+
+    private boolean isMember(String password, Member member) {
+        return member != null && member.validatePassword(password);
```
member 조회시 password까지 일치하는 것을 찾지 않고 validation해서 따로 처리하는 이유가 궁금합니당~~!
휴이의 의도가 있을까요?!

jh280722 replied:
member에서 get이랑 equals를 하는 것이 맞나 의문이 들어 이렇게 처리했습니다. 그냥 겟을 이용해서 한번에 처리하는게 좋을까요?

limmmee-bonjour commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,40 @@
+
+    public Member checkInvalidToken(String accessToken) {
+        if (!jwtTokenProvider.validateToken(accessToken)) {
+            throw new InvalidTokenException();
```
Exception 정의를 잘 해주셨어요 👏  그런데, 정확하게 어떤 메세지인지 사용자가 알기 어려울 것 같아요~
Exception을 던질때는 꼭 메세지와 함께 던져주세요!

limmmee-bonjour commented on src/main/java/subway/auth/ui/AuthController.java:
```diff
@@ -1,5 +1,42 @@
+        return ResponseEntity.ok().body(new TokenResponse(jwtTokenProvider.createToken(email)));
+    }
+
+    @ExceptionHandler(InvalidMemberException.class)
+    public ResponseEntity<Void> invalidToken() {
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+    }
```
Exception을 핸들링해서 Response를 넘겨주도록 작업해주셨네요 👍 👏

limmmee-bonjour commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,46 @@
+                    stationService.findStationById(favoriteRequest.getTarget()));
+            return FavoriteResponse.of(favoriteDao.insert(favorite));
+        }
+        return null;
```
null을 리턴하는 경우에는 항상 조심하는 것이 좋아요 :)
외부에서 NPE를 만날때를 대비해서 항상 처리를 해주어야하거든요!
명시적으로 Optional로 리턴하던지, null인 경우 Exception 처리를 해주던지 하는게 좋을 것 같아요!

limmmee-bonjour commented on src/main/java/subway/line/domain/Line.java:
```diff
@@ -1,36 +1,60 @@
         this.sections = sections;
     }
 
+    public static Line of(String name, String color) {
+        return new Line(name, color);
+    }
+
+    public static Line of(String name, String color, int extraFare) {
+        return new Line(name, color, extraFare);
+    }
+
+    public static Line of(Long id, String name, String color, int extraFare) {
+        return new Line(id, name, color, extraFare);
+    }
+
+    public static Line of(Long id, String name, String color, int extraFare, Sections sections) {
+        return new Line(id, name, color, extraFare, sections);
+    }
```
정적 팩토리 메서드로 변경해주셨네요 👍 💯

limmmee-bonjour commented on src/main/java/subway/line/domain/Sections.java:
```diff
@@ -2,17 +2,14 @@
 
     public List<Section> getSections() {
-        return sections;
+        return Collections.unmodifiableList(sections);
```
불변객체를 리턴하도록 해주었네요 👍

limmmee-bonjour commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,63 @@
+        this.lineService = lineService;
+    }
+
+    public PathResponse getShortPath(Long sourceId, Long targetId, AGE age) {
+        List<Line> lines = lineService.findLines();
+        DijkstraShortestPath dijkstraShortestPath = getDijkstraShortestPath(lines);
+        List<Vertex> vertexs = dijkstraShortestPath.getPath(Vertex.of(sourceId, DEFAULT_EXTRA_FARE), Vertex.of(targetId, DEFAULT_EXTRA_FARE)).getVertexList();
+
+        int distance = (int) dijkstraShortestPath.getPathWeight(Vertex.of(sourceId, DEFAULT_EXTRA_FARE), Vertex.of(targetId, DEFAULT_EXTRA_FARE));
+        int distanceFare = DistanceFare.getDistanceFare(distance);
+        int lineExtraFare = vertexs.stream()
+                .map(v -> v.getExtraFare())
+                .max(Integer::compare).orElse(DEFAULT_EXTRA_FARE);
+
+        return PathResponse.of(vertexs.stream()
+                .map(vertex -> StationResponse.of(stationService.findStationById(vertex.getStationId())))
+                .collect(Collectors.toList()), distance, calculateFare(age, distanceFare, lineExtraFare));
+    }
```
최단거리를 구하는 것은 중요한 요구사항인데, 테스트케이스가 없네요.
또 메소드 내에 너무 많은 로직이 담겨있어요 :) (테스트시에 PathResponse까지 만들어야하니 테스트가 어렵겠죠?)
물론 통합 테스트도 있지만, 단위테스트도 중요하답니다.
[도전!] 테스트케이스를 만들어주세요 (외에 중요한 로직들도 모두 테스트케이스를 만들어주세요)

limmmee-bonjour commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,40 @@
+        this.favoriteService = favoriteService;
+    }
+
+    @PostMapping("/favorites")
```
favorites URL이 중복되네요 :) 
클래스에 @RequestMapping 어노테이션을 사용해서 중복을 줄여봅시다 😉

limmmee-bonjour commented on src/main/java/subway/path/domain/Vertex.java:
```diff
@@ -0,0 +1,38 @@
+        return extraFare;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Vertex vertex = (Vertex) o;
+        return Objects.equals(stationId, vertex.stationId);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(stationId);
+    }
```
equals와 hasCode를 오버라이드 한 이유가 있을까요~?

jh280722 replied:
다익스트라에 vertex를 넣어서 경로를 구했는데 extraFare에 따라 다른 정점이 될 수 있기에 정점 역만 가지고 판단을 하기위해
equals에서 stationId로만 비교판단을 하기위해 사용했습니다. extraFare를 꺼내기 위해선 정점 클래스가 필요했고요

limmmee-bonjour commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,63 @@
+    public PathResponse getShortPath(Long sourceId, Long targetId, AGE age) {
+        List<Line> lines = lineService.findLines();
+        DijkstraShortestPath dijkstraShortestPath = getDijkstraShortestPath(lines);
+        List<Vertex> vertexs = dijkstraShortestPath.getPath(Vertex.of(sourceId, DEFAULT_EXTRA_FARE), Vertex.of(targetId, DEFAULT_EXTRA_FARE)).getVertexList();
```
Vertex에 기본요금이 셋팅되어있는 정적 팩토리 메서드를 만드는건 어떨까요~? 😉

jh280722 replied:
좋은 생각인 것 같아요!

limmmee-bonjour commented on src/main/java/subway/path/domain/DistanceFare.java:
```diff
@@ -0,0 +1,30 @@
+        if (distance <= TEN_KM.distance) {
+            return DEFAULT_FARE;
+        }
+        if (TEN_KM.distance < distance && distance <= FIFTY_KM.distance) {
```
TEN_KM.distance < distance 부분은 항상 참일 것 같아요! 
그리고 해당 메소드도 (getDistanceFare) 정말 중요한 요구사항인데, 테스트케이스가 없어요 :) 추가해주세요!

limmmee-bonjour commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,63 @@
+        List<Vertex> vertexs = dijkstraShortestPath.getPath(Vertex.of(sourceId, DEFAULT_EXTRA_FARE), Vertex.of(targetId, DEFAULT_EXTRA_FARE)).getVertexList();
+
+        int distance = (int) dijkstraShortestPath.getPathWeight(Vertex.of(sourceId, DEFAULT_EXTRA_FARE), Vertex.of(targetId, DEFAULT_EXTRA_FARE));
+        int distanceFare = DistanceFare.getDistanceFare(distance);
+        int lineExtraFare = vertexs.stream()
+                .map(v -> v.getExtraFare())
+                .max(Integer::compare).orElse(DEFAULT_EXTRA_FARE);
+
+        return PathResponse.of(vertexs.stream()
+                .map(vertex -> StationResponse.of(stationService.findStationById(vertex.getStationId())))
+                .collect(Collectors.toList()), distance, calculateFare(age, distanceFare, lineExtraFare));
```
현재 요금을 계산하는 정책이 "거리", "노선별", "연령별" 정책이 존재하네요.
그런데 해당 요금을 계산하는 로직이 distanceFare, lineExtraFare, calculateFare 여러 부분으로 나뉘어져 있어요.
만약 정책이 추가되면 해당 로직을 바꾸어야겠죠? calculateFare 메서드 파람은 점점 늘어날거예요.
게다가 각 정책은 정책별로 기능이 정리가 되어야할 것 같아요 :)
[도전!] 자 전략패턴을 사용해서 좀 더 유연한 구조로 바꿔 보아요! 
[이글](https://velog.io/@dltlsgh5/%EC%A0%84%EB%9E%B5-%ED%8C%A8%ED%84%B4strategy-pattern)이 도움이 될거예요 :)

jh280722 replied:
전략패턴을 사용한다면 인자가 모두 같아야 할 수 있지않나요?
distance와 lineFare는 다익스트라패스가 calculateFare부분에서는 age 인자가 필요한데 그걸 통합해서 할 수 있는 방법이 있을까요?

jh280722 replied:
감사합니다!

limmmee-bonjour commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.stream.Collectors;
+
+@Service
+public class FavoriteService {
```
Q. 도메인의 서비스에서 다른 도메인의 dao를 이용하고 싶을때 서비스를 불러와서 처리하는지 dao를 불러와서 처리하는지가 궁금합니다.
A. 위치는 여기는 아니지만 질문 주신 부분에 대해 제 의견을 드려요 :)
서로 다른 도메인의 DAO를 이용하고 싶을 수 있죠. 이 때는 다양하게 처리할 수 있습니다.
Service를 의존하기도하고 Dao를 의존하기도 합니다.
(약간 이건 팀 코드 스타일에 따라 다른 것 같습니다 :), 정답은 없습니다.)

service,dao던 어떤 것을 사용하는 것 보다는 어떻게 의존성을 연결할 것인가가 중요한 문제라고 생각해요.
순환참조라 던가 하는 의도치 않은 상황이 발생하기 때문에 항상 고민하고 의존성을 추가해주어야 합니다!
의존성이 어떻게 향하는지 항상 고민하고 호출을 해주어야 합니다 😉

jh280722 replied:
넵 감사합니다

limmmee-bonjour commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,31 @@
+
+    @GetMapping("/paths")
+    public ResponseEntity<PathResponse> getShortPath(@AuthenticationPrincipal LoginMember loginMember, @RequestParam("source") Long sourceId, @RequestParam("target") Long targetId) {
+        if (loginMember == null) {
+            return ResponseEntity.ok().body(pathService.getShortPath(sourceId, targetId, AGE.ADULT));
+        }
```
Q. path부분에서 로그인을 하지않아도 결과를 얻을 수 있고 로그인을 하면 나이에 따라 할인이 되어 출력이 되길래
resolveArgument에서 인증 토큰이 없을 경우는 null을 리턴하여 처리하였습니다.
이 부분이 맞게 한 것인지 궁금하고요.

A. 일단 요구사항을 보니 따로 해당 기능(로그인 안했을때)에 대해서는 언급이 없네요.
이렇게 구현해서 처리해주셔도 될 것 같아요.
그런데 만약에 나중에 LoginMember에 대한 유효성 체크를 하게 된다면, 그때는 해당 기능은 Controller에 있는 게 아니라
다른 로직에 있어야겠죠? 😉 
(현재 LoginMember가 꼭필요한 FavoriteController 에서도 따로 로그인에 대한 유효성은 없네요. 추가하면 좋을 것 같긴합니다 :))

jh280722 replied:
@AuthenticationPrincipal을 통해 resolveArgument에서 로그인 유효성을 판단하고 있습니다.

limmmee-bonjour reviewed (changes requested):
안녕하세요 huey! 이번 과제를 함께 하게 된 벨입니다 😄 
빠르게 잘 구현해주셨어요 👏 👏  (저는 리뷰가 좀 늦었네요^^;;)
잘 구현해주셨지만, 중요한 요구사항에 테스트들이 모두 빠져있네요.
물론 통합테스트도 중요하지만, 단위테스트들도 꼭 챙겨주셔야합니다!
테스트케이스 추가 및 요금 계산하는 부분에 조금 더 유연하게 처리할 수 있지 않을까싶어 의견드렸어요 :)
[도전!] 이라고 남긴 부분은 고민하셔서 수정해주시면 좋을 것 같아요.
궁금하신 점이 있다면 언제든 DM 주셔요! 그럼 화이팅입니다!

Commit:
refactor : 예외 메세지 출력 및 전반적인 리펙토링

Commit:
refactor : 요금 전략패턴 적용

Commit:
refactor : 정적 팩토리 메서드로 변경

Commit:
refactor : DistanceFare enum 제거 후 distanceFare 전략패턴으로 통합

Commit:
refactor : Lines 구현 및 다익스트라 관련 역할 이전

Commit:
test : 거리, 나이, 라인 관련 요금 테스트

limmmee-bonjour commented on src/main/java/subway/auth/exception/InvalidMemberException.java:
```diff
@@ -1,4 +1,7 @@
 package subway.auth.exception;
 
 public class InvalidMemberException extends IllegalArgumentException {
+    public InvalidMemberException(String message) {
```
Exception 메세지 정의 좋습니다 👍

limmmee-bonjour commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -16,22 +16,21 @@
+        if (favoriteRequest.getSource() == null || favoriteRequest.getTarget() == null) {
+            throw new InvalidFavoriteException("즐겨찾기 요청 정보가 비어있습니다.");
         }
-        return null;
```
null을 리턴하는 대신 Exception 처리해주셨군요 👏

limmmee-bonjour commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,40 +1,45 @@
 import java.util.List;
 
 @RestController
+@RequestMapping("/favorites")
```
RequsetMapping 사용 굿 👍

limmmee-bonjour commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -34,18 +35,23 @@ public PathResponse getShortPath(Long sourceId, Long targetId, AGE age) {
         List<Vertex> vertexs = dijkstraShortestPath.getPath(Vertex.of(sourceId), Vertex.of(targetId)).getVertexList();
```
이전 리뷰시에 [도전!] 으로 getShortPath의 테스트케이스가 누락되어 포함해 달라고 요청드렸는데, 
이번 리뷰에 반영이 안되어있네요! 혹시 6개의 리뷰가 hidden 처리되어 확인을 못하신건지 누락된건지 다시 한번 확인 부탁드려욥!
[도전!] 중요한 요구사항들은 단위테스트를 구현해주세요

jh280722 replied:
pathService에서의 테스트는 PathAcceptanceTest에서 테스트케이스가 등록되어있어서 생략했습니다. 어떤 점을 더 테스트를 해주는 게 좋을까요? 단위 테스트로 쪼개서 distance, stations, fare를 독립적으로 테스트를 하였고 전체 메서드를 실행하는 테스트는 pathAcceptanceTest에 있는데 제가 테스트들을 커밋하지 않아서 못 보신 것 같습니다. 대신 각각의 요금, 거리, 스테이션에 해당되는 테스틑 추가했습니다.

limmmee-bonjour replied:
물론 pathAcceptanceTest라는 통합 테스트가 있지만, 통합테스트 말고도 해당 메서드의 유닛테스트가 필요해보여 추가드린거였어요.

제가 생각했을때 getShortestPath 메서드는 path를 구하는 요구사항에서 중요한 메서드로 보여집니다.
최단거리의 메서드를 가져오는 로직이 이 안에 집약되어 있기 때문인데요.
함께 개발하는 개발자가 해당 코드를 봤을때 최단거리의 station, 거리, 요금을 한 메서드에서 구현하고 있어서 한눈에 알아보기 어려울 것 같아요. 해당 메서드의 단위테스트를 구현하면, 테스트케이스에 어떠어떠한 경우에 해당 로직이 최단거리를 나타내는지에 대해 정의가 되어있어좀 더 이해도가 높아질 것 같습니다. 물론 여러 케이스에 대해 테스트해볼 수도 있구요.

뿐만 아니라 현재 pathAcceptanceTest 테스트는 /paths API를 가진 통합테스트로 테스트를 하는데, 해당 API에 다른 기능이 추가된다면
온전히 가장짧은 거리에 대한 테스트를 할 수 있을까요? 통합테스트도 중요하지만, 해당 메서드별 단위테스트를 만드는것 또한 중요하다고 생각합니다.

limmmee-bonjour commented on src/main/java/subway/path/domain/strategy/FareStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package subway.path.domain.strategy;
+
+public interface FareStrategy {
```
전략 패턴을 사용하여 요금을 계산하도록 해주셨네요 👍 
금액 계산이 각 패턴에 들어가서 유연해지고 가독성도 높아졌어요

limmmee-bonjour commented on src/main/java/subway/path/domain/strategy/ageFare.java:
```diff
@@ -0,0 +1,15 @@
+    private final AGE age;
+    public ageFare(AGE age) {
+        this.age = age;
+    }
```
각각의 구현인터페이스에 param이 다르기 때문에 해당 인스턴스 변수를 상태로 지정해주신 걸까요~?
금액 계산하는 distance, age, vertexs 는 value로 만들어서 해당 value를 param으로 넘길 수 있도록 하면 어떨까요?

물론 객체 지향으로 생각했을 때  distance, age, vertexs가 묶여 추상적인 의미를 가질 수 있는지 고민해봐야합니다.
아무거나 막 묶어서 param으로 넘기면 안되니까요.
만약 이 3개가 묶일 수 있다면 어떤 이름으로 어떤 행동을 가지고 묶일 수 있을지 추상화 할 수 있을지 고민해보아요!

limmmee-bonjour commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -34,18 +35,23 @@ public PathResponse getShortPath(Long sourceId, Long targetId, AGE age) {
+        fare = fareStrategy.apply(fare);
+
+        fareStrategy = new ageFare(age);
+        fare = fareStrategy.apply(fare);
```
fare를 계산하는 부분은 service가 아닌 다른 곳에서 처리할 수 있을 것 같아요.
DEFAULT_FARE도 service에서는 딱히 관심사가 없을 것 같고요! :)

[도전] 이를테면 fareCreator? 같이 요금 생성에 관련된 객체 or 팩토리 클래스를 만들어봅시다!

그리고 현재 구조는 전략패턴을 사용했는데 조금 아쉬운 것 같아요.
전략이 늘어날때마다 새로운 구현체를 추가해줘야 하네요. 이건 어때요?

실제 인터페이스에서 apply 메소드를 계산해서 주고 있는데,
기본 금액에 -, + 금액만 계산하도록 하면 어떨까요? 그리고 마지막에는 기본 금액에서 sum 만 해주는거죠!
ex) 기본 요금 + (거리에따른 추가금액) + (나이에따른 추가/감소금액) -> 1250 + (100) + (-500) 

만약 요금전략에 포함되지 않으면 그냥 filter에서 빠지게 해도 되고요 (이건 들어가도 되고 안들어가도 됩니다)
아래처럼 된다면 new 해서 구현클래스를 정의해주는 것보다 유연해지지 않을까요? 😉 
```
int extraFare = strategies.stream()
        .filter(strategy -> strategy.filter(createValue))
        .mapToInt(strategy -> strategy.discount(createValue))
        .sum();
return DEFAULT_FARE + extraFare;
```

물론 정답은 없습니다 :) 더 유연한 로직을 만들수있게 고민해보아요~

jh280722 replied:
좋은 방법인 것 같아요! 근데 ageFare의 경우 나머지 요금들의 합의 비율로 값이 정해지는데 그러면 결국 다 더하는 방법으로는 안되지 않을까요?

jh280722 replied:
팩토리 클래스를 만들어서 거기서 처리를 해주게 해볼게요

limmmee-bonjour commented on src/test/java/subway/path/FareTest.java:
```diff
@@ -1,4 +1,85 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
 public class FareTest {
```
FareTest에 한번에 전략에 관련 테스트를 모아보기보다는 구현클래스Test를 만들어서 그 클래스에 해당 전략에 대한 테스트를 만드는게 좋습니다!
[ 도전! ] 각 구현클래스Test를 만들어서 각 클래스를 옮겨주세요

limmmee-bonjour commented on src/test/java/subway/path/FareTest.java:
```diff
@@ -1,4 +1,85 @@
+    @DisplayName("나이 할인 요금 테스트")
+    @Test
+    void ageFare() {
+        fareStrategy = new AgeFare(AGE.BABY);
+        assertThat(fareStrategy.apply(1350)).isEqualTo(0);
+
+        fareStrategy = new AgeFare(AGE.CHILD);
+        assertThat(fareStrategy.apply(1350)).isEqualTo(850);
+
+        fareStrategy = new AgeFare(AGE.TEENAGER);
+        assertThat(fareStrategy.apply(1350)).isEqualTo(1150);
+
+        fareStrategy = new AgeFare(AGE.ADULT);
+        assertThat(fareStrategy.apply(1350)).isEqualTo(1350);
```
동일한 테스트 패턴이 계속되는 것 같아요 :)
Junit5에는 굉장히 다양한 테스트 어노테이션들이 추가되었습니다!
[해당 글](https://www.baeldung.com/parameterized-tests-junit-5)을 읽어보시면 추후 노가다? 없이 (ㅎㅎ) 간결한 테스트 케이스를 작성하실 수 있으실 거예요!

limmmee-bonjour reviewed (changes requested):
안녕하세요 휴이! 벨입니다 😄 
빠르게 수정하여 리뷰 주셨네요 👍  
다만 이전에 [도전!]으로 드린 부분이 있는데, 리뷰가 접혀져서 있어서 그런지 확인을 못하신 것 같아요
제가 드린 리뷰 중 반영 안된 부분 다시 확인 드립니다. 요청드린 테스트케이스는 꼭 추가 부탁드리고요 :)
전략패턴을 잘 써주셨는데👏 , 좀 더 유연하게 처리될 수 있도록 코멘트 드렸어요
확인해보시고 혹시 궁금하신 점이 있다면 언제든지 DM 주셔요 :) 화이팅입니다!

Commit:
refactor : parameterizedTest로 변경

Commit:
refactor

- 전략패턴 팩토리 클래스 구현 및 stream을 통한 extraFare 합 구현
- ageFare 제거 및 age enum에 역할 부여
- 패키지 구조 변경
- Vertices를 통한 일급 컬렉션 구현

Commit:
test : Vertices 및 통합 요금 테스트 추가

limmmee-bonjour commented on src/main/java/subway/member/domain/Age.java:
```diff
@@ -1,6 +1,6 @@
 package subway.member.domain;
 
-public enum AGE {
+public enum Age {
```
오호 저도 이건 미처못봤는데! 잘 변경해주셨네요 👏

limmmee-bonjour commented on src/test/java/subway/member/AgeTest.java:
```diff
@@ -0,0 +1,47 @@
+public class AgeTest {
+    @DisplayName("나이 할인 요금 테스트")
+    @ParameterizedTest
+    @MethodSource("ageFareParameter")
```
테스트 케이스를 전략에 맞게 잘 구현해주셨네요 !
요렇게 처리할수도 있어요!

```
@CsvSource(value = {
            "10 | 1250",
            "11 | 1350"
    }, delimiter = '|')
```

limmmee-bonjour commented on src/main/java/subway/path/domain/fare/strategy/FareStrategyFactory.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class FareStrategyFactory {
```
factory를 구현해주셨네요 :) ! 👍 

Q. 근데 ageFare의 경우 나머지 요금들의 합의 비율로 값이 정해지는데 그러면 결국 다 더하는 방법으로는 안되지 않을까요?
A. 네네 만약 그게 안된다면, [데코레이터 패턴](https://gdtbgl93.tistory.com/9)으로 고민해보는 방법도 좋을 것 같습니다~
끊임없이 더 나은 방향을 고민하면 좋을것 같아요 :)

jh280722 replied:
오 그렇군요 몰랐던 패턴인데 공부해보겠습니다. 감사합니다!

limmmee-bonjour reviewed (approved):
안녕하세요 휴이! 벨입니다.
요청 드린 요구사항들을 잘 구현해주셨네요! 
단위테스트 관련한 의견 및 소소한 의견 달아뒀으니 확인해주세요!
항상 답은 없으니 휴이만의 답을 찾아가면 될 것 같습니다. 해당 미션 진행하시느라 고생 많으셨습니다.
궁금하신 부분이 있으시면 언제든지 DM 주시기 바랍니다.
